### PR TITLE
 UI for Special Resolution Form Details#12926

### PR DESCRIPTION
**Pull request by shabeeb-aot**
*Issue #:* /bcgov/entity#12926

*Description of changes:*
* Special resolution form 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Comment by shabeeb-aot** [comment]
I am adding will add unit test case and validation for this component

File: src/interfaces/special-resolution-interfaces/create-resolution-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/create-resolution-interface.ts:9]: It would be helpful to document, beside each date (lines 6 and 9), what format they are, eg, `// YYYY-MM-DD`.

File: src/interfaces/special-resolution-interfaces/person-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/person-interface.ts:1]: Please rename this (and EmptyPerson below) to distinguish them from OrgPersonIF and ApiPersonIF.

File: src/interfaces/special-resolution-interfaces/person-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/person-interface.ts:11]: The empty values should be the same as those returned by the API when a person object is fetched.

For example, will all of theses be empty strings? Or could they be null?

Here's what I'm trying to avoid for you: https://github.com/bcgov/business-edit-ui/blob/e7f3877a3f870a0ddba63c9524f8f97073614829/src/store/getters/state-getters.ts#L728

Comment (severinbeauvais) [reply src/interfaces/special-resolution-interfaces/person-interface.ts:11]: Maybe the optional fields should be null?

Or maybe additionalName shouldn't be optional?

File: src/interfaces/store-interfaces/state-model-interface.ts
Comment (severinbeauvais) [inline src/interfaces/store-interfaces/state-model-interface.ts:42]: Let's get away from calling these "steps" (which is left over from Create UI).

This is just a data object, and could be reused by any filing.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:4]: Duplicate v-card.

Also please watch your whitespace and indents.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:38]: getCreateResolutionResource should never be falsy right?

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:195]: Delete.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:213]: This is a short line. Move it all to the same line please.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:232]: This should be
`readonly MAX_RESOLUTION_TEXT_LENGTH = 1000`

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:230]: Some versions of lint complain that you don't need to declare it's a string when it's already initialized as a string, so I take that as a best practice :)

-> Local variables should not have an explicit type when they're initialized to something with a type.

Examples:
```
protected myString = ``
protected myNumber = 123
protected myObj = null as myObjIF
```

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:242]: How about falling back to `{}` so that you can still de-reference it in the template (and remove the check if falsy)?

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:259]: Why don't you just hard-code the path in the img tag [like this](https://github.com/bcgov/business-create-ui/blob/73cd0b9344031b50579ff7bb898551c8adf0df17/src/views/Incorporation/IncorporationAgreement.vue#L196)?

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:269]: Don't use `new Date()`. Ever. It creates a date object based on the user's browser and workstation settings.

Instead, in this case, use the Date Mixin methods to convert an "api date" to a JS date object.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:278]: Again, don't do this. You don't know what date you'll get.

There's a getter for current date (yyyy-mm-dd) already. (And it is updated periodically in the background.)

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:292]: As above.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:317]: These 3 "new Dates" are actually OK since you're only using them for comparison.

However, I think you could just compare the dates (yyyy-mm-dd), eg.
"is 2022-08-01 < 2022-08-02"? yes
"is 2022-08-01 > 2022-08-02"? no

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:321]: Add a comment why you need this.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:337]: Should be private.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:347]: Don't do this. Declare all class variables at the top with the others. Also should not need semi-colons.

Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:347]: Maybe these rules can be readonly also?

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:360]: No. Use getCurrentDate.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:374]: No.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:392]: Add a comment why you need this.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:402]: Add a comment why you need this.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:409]: Created should be protected, and our code style is to place it near the top of the class (possibly just under the local getters, which are kind of like variables).

Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:409]: Now that I look at what this does, can't you do it during the variable initializations?

If you depend on the getter, note that these will be assigned the getter values as of when this component is created. They will not be assigned a subsequent getter value (not reactive).

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:268]: `getCurrentDate` is not in API format. Use a different method to convert yyyy-mm-dd to a Date.

Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:268]: Also on lines 341 and 351.

File: src/interfaces/special-resolution-interfaces/create-resolution-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/create-resolution-interface.ts:3]: Since you are reusing this IF (hence its name from Create UI), please add a comment that this IF is also using for Special Resolution filings.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [inline src/components/SpecialResolution/SpecialResolutionForm.vue:238]: This is used by HelpSection component. If that component does not already handle falsy (undefined) then you should have a fallback here (like line 242 maybe?).

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/signing-person-interface.ts:12]: Additional Name is optional but null.
Email is also optional. Should it be initialized here?

File: src/resources/SpecialResolution/CooperativeResource.ts
Comment (severinbeauvais) [inline src/resources/SpecialResolution/CooperativeResource.ts:28]: Should be "If you require".

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (severinbeauvais) [inline tests/unit/SpecialResolutionForm.spec.ts:30]: This may not work in GH because it will have a different date library / timezone than you. If so then you will have to mock this getter.

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (severinbeauvais) [inline tests/unit/SpecialResolutionForm.spec.ts:35]: It's better to hard-code data here instead of using IFs or objects, so that 2 mistakes don't cancel each other out.

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (severinbeauvais) [inline tests/unit/SpecialResolutionForm.spec.ts:67]: You should be able to await the trigger function (previous line) and not need this line.

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (severinbeauvais) [inline src/interfaces/special-resolution-interfaces/signing-person-interface.ts:8]: Don't name this xxxIF. It's not an IF. It's just EmptySigningPerson (object).

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (shabeeb-aot) [reply src/interfaces/special-resolution-interfaces/signing-person-interface.ts:12]: Additional Name if we didn't define null have to put additional check html
we don't have an email in design

File: src/resources/SpecialResolution/CooperativeResource.ts
Comment (shabeeb-aot) [reply src/resources/SpecialResolution/CooperativeResource.ts:28]: this is the same for SP/GP also. I copied it from Sp/GP resource file. Can I change all places?

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (severinbeauvais) [reply src/interfaces/special-resolution-interfaces/signing-person-interface.ts:12]: I'm looking for consistency between the interface and this empty object. Why are additional name and email different? I'm thinking they should be declared and initialized the same...?

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (shabeeb-aot) [reply tests/unit/SpecialResolutionForm.spec.ts:72]: it fails without await. since its a click function

File: src/resources/SpecialResolution/CooperativeResource.ts
Comment (severinbeauvais) [reply src/resources/SpecialResolution/CooperativeResource.ts:28]: Oops! Yes, please fix all 3 :)

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (shabeeb-aot) [reply src/interfaces/special-resolution-interfaces/signing-person-interface.ts:11]: my bad, email is not there removed from interface

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (severinbeauvais) [reply tests/unit/SpecialResolutionForm.spec.ts:72]: Weird. It works in other unit tests in this project.

Maybe it's because you need to wait for some async validation to complete before you can verify the result. If this is so, please add a comment here that this is why the "await nextick" is needed.

You should also await the trigger anyway since it's async.

File: src/interfaces/special-resolution-interfaces/signing-person-interface.ts
Comment (severinbeauvais) [reply src/interfaces/special-resolution-interfaces/signing-person-interface.ts:11]: Bye bye, email.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (shabeeb-aot) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:303]: you mean to compare strings?

Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:303]: Yes, since they're all yyyy-mm-dd format, right?

Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:303]: I leave this up to you.

**Review by severinbeauvais — APPROVED** [review]
I'll leave the last decisions/cleanup up to you. Looks good otherwise.

**Comment by severinbeauvais** [comment]
By the way, for future PR reviews, I will wait for the other reviewers to look at it before I do. This will give others a chance to comment and should reduce the load on me.

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (shabeeb-aot) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:303]: As far as I heard it's better to compare with a date than string.

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (shabeeb-aot) [reply tests/unit/SpecialResolutionForm.spec.ts:35]: I am not using this value to compare now, I just used it to render correctly, So the test will not fail in this case, I will revisit if needed

**Review by lambert-alex — APPROVED** [review]
LGTM

File: tests/unit/SpecialResolutionForm.spec.ts
Comment (lambert-alex) [inline tests/unit/SpecialResolutionForm.spec.ts:72]: Not important but should be until not untill

**Comment by lewischenstudio** [comment]
Everything looks good, but the `package.json` file didn't update the app version. I'll approve this PR once this is updated.

**Review by lewischenstudio — APPROVED** [review]

File: src/components/SpecialResolution/SpecialResolutionForm.vue
Comment (severinbeauvais) [reply src/components/SpecialResolution/SpecialResolutionForm.vue:303]: OK.

**Pull request merged**
